cirq-core/cirq/transformers/optimize_for_target_gateset.py

```diff
@@ -126,20 +135,32 @@ def optimize_for_target_gateset(
         return _decompose_operations_to_target_gateset(
             circuit, context=context, ignore_failures=ignore_failures
         )
-
-    for transformer in gateset.preprocess_transformers:
-        circuit = transformer(circuit, context=context)
-
-    circuit = _decompose_operations_to_target_gateset(
-        circuit,
-        context=context,
-        gateset=gateset,
-        decomposer=gateset.decompose_to_target_gateset,
-        ignore_failures=ignore_failures,
-        tags_to_decompose=(gateset._intermediate_result_tag,),
-    )
-
-    for transformer in gateset.postprocess_transformers:
-        circuit = transformer(circuit, context=context)
-
+    if isinstance(max_num_passes, int):
+        _outerloop = lambda: range(max_num_passes)
+    else:
+
+        def _outerloop():
+            while True:
+                yield 0
+
+    initial_num_moments, initial_num_ops = len(circuit), len(tuple(circuit.all_operations()))
+    for _ in _outerloop():
+        for transformer in gateset.preprocess_transformers:
+            circuit = transformer(circuit, context=context)
+        circuit = _decompose_operations_to_target_gateset(
+            circuit,
+            context=context,
+            gateset=gateset,
+            decomposer=gateset.decompose_to_target_gateset,
+            ignore_failures=ignore_failures,
+            tags_to_decompose=(gateset._intermediate_result_tag,),
+        )
+        for transformer in gateset.postprocess_transformers:
+            circuit = transformer(circuit, context=context)
+
+        num_moments, num_ops = len(circuit), len(tuple(circuit.all_operations()))
```

Comment:
No need to create an intermediate tuple. 

```suggestion
        num_moments, num_ops = len(circuit), sum(1 for _ in circuit.all_operations())
```

Reply:
done